app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt

```diff
@@ -0,0 +1,1166 @@
+    hikeFlow.value = newHikeState
+
+    // Update the selected hike if necessary
+    updateSelectedHike()
+
+    return true
+  }
+
+  private suspend fun unsaveHikeAsync(
+      hikeId: String,
+      onSuccess: () -> Unit,
+      onFailure: () -> Unit
+  ) =
+      withContext(dispatcher) {
+        var successful = false
+        _hikesMutex.withLock {
+          // Retrieve the hike to unsave from the loaded hikes
+          val hikeFlow = _hikeFlowsMap[hikeId] ?: return@withLock
+          val hike = hikeFlow.value
+
+          // Avoid updating the hike if it is not saved
+          if (!hike.isSaved) {
+            successful = true
+            return@withLock
+          }
+
+          // Unsave the hike using the repository
+          try {
+            savedHikesRepo.removeSavedHike(SavedHike(hike.id, hike.name ?: "", hike.plannedDate))
+          } catch (e: Exception) {
+            Log.e(LOG_TAG, "Error encountered while unsaving hike", e)
+            successful = false
+            return@withLock
+          }
+
+          // Update the saved hikes map
+          _savedHikesMap.remove(hikeId)
+
+          if (_loadedHikesType == LoadedHikes.FromSaved) {
+            // Only saved hikes may stay in the list, delete the unsaved hike from the list
+            _hikeFlowsMap.remove(hikeId)
+            updateHikeFlowsList()
+          } else {
+            // The hike can stay even if it is not saved, so update it
+            val newHikeState = hikeFlow.value.copy(isSaved = false, plannedDate = null)
+            hikeFlow.value = newHikeState
+          }
+
+          // Update the selected hike if necessary
+          updateSelectedHike()
+
+          successful = true
+        }
+
+        // Perform the callback without the lock, to avoid deadlocks and improve performance
+        if (successful) {
+          onSuccess()
+        } else {
+          onFailure()
+        }
+      }
+
+  private suspend fun setPlannedDateAsync(
+      hikeId: String,
+      date: Timestamp?,
+      onSuccess: () -> Unit,
+      onFailure: () -> Unit
+  ) =
+      withContext(dispatcher) {
+        var successful = false
+        _hikesMutex.withLock {
+          // Retrieve the hike to update from the loaded hikes
+          val hikeFlow = _hikeFlowsMap[hikeId] ?: return@withLock
+          successful = updateHikePlannedDateAsync(hikeFlow, date)
+        }
+
+        if (successful) {
+          onSuccess()
+        } else {
+          onFailure()
+        }
+      }
+
+  /**
+   * Helper function to update the planned date of a hike.
+   *
+   * This function requires [_hikesMutex]'s lock to be acquired before being called. It is the
+   * caller's responsibility to acquire the lock.
+   *
+   * @param hikeFlow The flow of the hike to update.
+   * @return True if the operation was successful, false otherwise.
+   */
+  private suspend fun updateHikePlannedDateAsync(
+      hikeFlow: MutableStateFlow<Hike>,
+      date: Timestamp?
+  ): Boolean {
```

Comment:
The only reason was to have this part :

```kotlin
val hikeFlow = _hikeFlowsMap[hikeId] ?: return@withLock
```

Before calling that method, and providing it directly the hike flow to avoid searching for it again.

I could move it but it's an internal method so I'm not sure it is necessary. What do you think?

Reply:
True, this is completely unnecessary and it makes more sense to have the same way of doing everywhere. Changed it 👌🏻

Reply:
Thanks for the change!